rust_src/src/eval.rs

```diff
@@ -1338,4 +1338,143 @@ pub fn condition_case(args: LispCons) -> LispObject {
     unsafe { internal_lisp_condition_case(var, bodyform, handlers) }
 }
 
+type SpecbindingRef = ExternalPtr<specbinding>;
+
+#[no_mangle]
+pub extern "C" fn specpdl_symbol(pdl: SpecbindingRef) -> LispObject {
+    debug_assert!(pdl.kind() >= SPECPDL_LET);
+    unsafe { pdl.let_.as_ref().symbol }
+}
+
+#[no_mangle]
+pub extern "C" fn specpdl_old_value(pdl: SpecbindingRef) -> LispObject {
+    debug_assert!(pdl.kind() >= SPECPDL_LET);
+    unsafe { pdl.let_.as_ref().old_value }
+}
+
+#[no_mangle]
+pub extern "C" fn set_specpdl_old_value(mut pdl: SpecbindingRef, val: LispObject) {
+    debug_assert!(pdl.kind() >= SPECPDL_LET);
+    unsafe {
+        pdl.let_.as_mut().old_value = val;
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn default_toplevel_binding(symbol: LispObject) -> SpecbindingRef {
+    let current_thread = ThreadState::current_thread();
+    let specpdl = SpecbindingRef::new(current_thread.m_specpdl);
+
+    let mut binding = SpecbindingRef::new(std::ptr::null_mut());
+    let mut pdl = SpecbindingRef::new(current_thread.m_specpdl_ptr);
+
+    while pdl > specpdl {
+        unsafe {
+            pdl.ptr_sub(1);
+        }
+        match pdl.kind() {
+            ref x if [SPECPDL_LET_DEFAULT, SPECPDL_LET].contains(x) => {
```

Comment:
Why use a guard here instead of `SPECPDL_LET_DEFAULT | SPECPDL_LET => { ... }`?

Reply:
I actually don't remember, but now I see that you're right. Fixed